### ISSUE TITLE
Runtime configuration for features

### ISSUE THREAD

**Issue by mathis-m** [swagger-api/swagger-ui#7033]
I would like to enable / disable given features that can be configured via the Swagger-UI configuration at runtime.

This would enable users to get to know several features.

Example if one does not know that there is a filter how he will never know because it is hidden behind a flag.

I want to highlight that with this, features, that cannot be enabled on default, because this would result in a breaking change, at least could be promoted via a given runtime configuration panel.

https://user-images.githubusercontent.com/11584315/110026607-20144180-7d31-11eb-87a2-07b373cfd4e2.mp4

**Comment by hkosova** [comment]
Looks interesting. I think this can be useful in the https://petstore.swagger.io demo. Or as part of a separate app that could generate Swagger UI config code based on the selected features, something like
https://prismjs.com/download.html#themes=prism&languages=markup+css+clike+javascript

But in production instances of hosted API docs, it is up to the docs _owner_ (not the users) to configure the apearance and available features for their Swagger UI instance.

**Comment by mathis-m** [comment]
@hkosova I totally agree for Embedded Swagger-UI instances. But this feature is is aimed at Swagger-UI being used as independent tool, for example out of the box integration into .Net 5 web api.

In addition we could give the Docs Owner the possibility to set the feature set they like. Everything that is set by organization // owner can not be changed by a user. Similar to the Azure DevOps Preview Features tab.  In addition users can activate additional features.

**Comment by mathis-m** [comment]
In addition I want to underline that users might have different needs when it comes to features.
A non technical user might not need to skip the Try Out but a developer does.

So how will the owner decide. Will he enable the Try Out by default to speed up developers and slow down non technical users or vice-versa. Why not leave the choice to the user.

**Comment by mathis-m** [comment]
One thing that should be taken into consideration is that swagger-ui is used as out of the box client and documentation. For this use case it would be cool if you can configure swagger-ui without having to create a customized version that possibly needs to be maintained.

**Comment by tim-lai** [comment]
I think this is a good feature. I think that providing an easy mechanism to customize the experience is a good thing. For most people, I think a GUI interface for options would be much easier to use than query parameters. Given that there are cases where the tool owner may limit selectable options by an individual end user, there should also be a mechanism for the tool owner to configure allowable individual options, including disabling of specific query parameters.

Considering the scope and stakeholder impact, I think this feature warrants more discussion and feedback.

**Comment by mathis-m** [comment]
For now I will go ahead and add the option to add a mask for features that are considered unchangeable.
Since all queryparameters are mapped into the features plugin, the mask will also apply to those.

**Comment by mathis-m** [comment]
I think we should also expose some action to enable / disable the popup state. So it can be integrated from anywhere.

**Comment by mathis-m** [comment]
@tim-lai would it make sense architectural wise to not just store the enabled flag, but to store the whole configuration of a feature.
So that each feature could implement its own configuration component (on default it could just use the toggle component).

**Comment by mathis-m** [comment]
@tim-lai some unrelated question when will the https://petstore3.swagger.io/ version of swagger-ui be updated?

**Comment by tim-lai** [comment]
> @tim-lai would it make sense architectural wise to not just store the enabled flag, but to store the whole configuration of a feature.
> So that each feature could implement its own configuration component if needed(on default it could just use the toggle component). Eg. Request Snippets > defaultExpanded could then be also toggled from the configuration panel.

Yes. 😸

**Comment by tim-lai** [comment]
> @tim-lai some unrelated question when will the https://petstore3.swagger.io/ version of swagger-ui be updated?

good question. I've submitted a ticket for to get petstore3 updated.

**Comment by tim-lai** [comment]
> I think we should also expose some action to enable / disable the popup state. So it can be integrated from anywhere.

I agree.

**Comment by mathis-m** [comment]
@tim-lai I have done some changes could you verify if you would like how I integrated it in #7034.

if you are fine with the architectural decisions I have taken in order to provide a unified way to integrate user config that is merged with organizational config, I will proceed to write some tests and add the popup actions. It also allows you to configure some initial settings that will be mounted into the global state on creation (in the according feature namespaces). I have moved requestSnippets to the new configuration point to showcase this.

**Comment by tim-lai** [comment]
@mathis-m The core changes in #7034 looks good to me, and good to move forward with iteration and tests. Once all working and passing tests, it would be helpful to update documentation that  addresses what is/is not available via runtime configuration, and if applicable, any order on precedence and conflicting legacy options.
